### DISCUSSION TITLE
How to find out if an font support an character?

### DISCUSSION THREAD

**Discussion by moi15moi** [fonttools/fonttools#2997] [Q&A] [answered]
How can I find out if an font support an character?

For example, I wanna know if this [font.zip](https://github.com/fonttools/fonttools/files/10775522/font.zip) support the glyph ``ë``.
In fontforge, I can clearly see that it support it.

I have try this, but it doesn't work:
```py
from fontTools.ttLib.ttFont import TTFont

def is_glyph_supported(font: TTFont, glyph: str):
    if len(glyph) != 1:
        raise Exception("Glyph need to be 1 character.")


    for table in font["cmap"].tables:
        if ord(glyph) in table.cmap:
            return True

    return False

font = TTFont("test.ttf")
print(is_glyph_supported(font, "ë")) # it return me false when it should return true
```

I have also try this, but it also doesn't work:
```py
from fontTools.ttLib.ttFont import TTFont
from fontTools.misc.encodingTools import getEncoding

def is_glyph_supported(font: TTFont, glyph: str):
    if len(glyph) != 1:
        raise Exception("Glyph need to be 1 character.")


    for table in font["cmap"].tables:
        if (glyph).encode(getEncoding(table.platformID, table.platEncID, table.language)) in table.cmap:
            return True

    return False

font = TTFont("test.ttf")
print(is_glyph_supported(font, "ë")) # it return me false when it should return true
```

**Comment by typoman** [comment]
You can get the character map this way:
```py
from fontTools.ttLib import TTFont

f = TTFont(font_path)
cmap = f.getBestCmap()
```

**Reply by moi15moi** [reply]
Like you can see in my question, i am already able to get the cmap.

I wanna know how i can find if the cmap contain x glyph

**Reply by moyogo** [reply]
What OS and what version of python are you using? There may be an encoding issue.

`ord(c) in font.getBestCmap()` works with python3.

**Reply by moi15moi** [reply]
I think fonttools may have a bug with the font I share.

Because, I tried with other font where the cmap isn't from the mac platform and it work properly.

OS: Windows 10
Python version: 3.10.7
Fonttools version: 4.38.0

**Reply by justvanrossum** [reply]
Looking at the font: it contains a single cmap subtable:

    <cmap_format_0 platformID="1" platEncID="0" language="0">

This is a cmap for a mac-specific encoding, and fonttools does not know how to translate this to unicode. The keys in the `.cmap` attributes correspond to the relevant mac encoding code points, not unicode code points.

**Reply by moi15moi** [reply]
I thought about that.

But, fonttools say that the character ``ë`` has the code ``0x98``.

But, the [mac roman table](https://en.wikipedia.org/wiki/Mac_OS_Roman) say that ``ë`` has the code ``0x91``.
And fontforge consider the character ``ë`` has the code ``0x91``.

**Reply by justvanrossum** [reply]
Ok, I figured it out. The encoding is indeed MacRoman, but the `post` table is incorrect, so fonttools doesn't get meaningful glyph names. 

- ë is indeed 0x91 in MacRoman
- in the cmap, 0x91 is mapped to the glyph "adieresis", because that's what the post table tells it to be
- If you *render* "adieresis", it actually *is* the glyph for "ë"

It's a font from 1992/1993 that was very poorly engineered and/or poorly converted. The copy from this source is better: https://www.abstractfonts.com/font/795 At least it has a good post table and a Unicode cmap subtable.

**Reply by moi15moi** [reply]
> If you render "adieresis", it actually is the glyph for "ë".

Then, why the glyph ``ë`` is properly render by GDI on Windows (and fontforge properly decoded) ?

**Reply by justvanrossum** [reply]
Above I describe how ë is rendered correctly *despite* it having a "wrong" glyph name. The glyph name is irrelevant for rendering. The cmap maps the MacRoman character to the correct glyph id. But FontTools maps the glyph id to the wrong name because of the bad post table. Change the post table to a format-3 post table and fonttools will make up glyph names based on the cmap. 

GDI does not use glyph names, and FontForge perhaps has a different way to make up glyph names, or also ignores them. FontTools relies on the post table to extract glyph names. The post table is wrong in this font. Therefore FontTools produces the "wrong" glyph names. It has no way to know that the post table is wrong, though.

**Reply by moi15moi** [reply]
> GDI does not use glyph names

Do you know how GDI map the character to an glyph?
I should have mentionned previously, but my goal is to reproduce how GDI map an caracter (like ``ë``) to the glyph.

**Reply by anthrotype** [reply]
the cmap is the only thing that maps a caracter to a glyph

**Reply by anthrotype** [reply]
> FontTools maps the glyph id to the wrong name because of the bad post table. Change the post table to a format-3 post table and fonttools will make up glyph names based on the cmap.

that's the problem, the buggy post table. Discard the glyph names altogether, by setting post to format 3.0 before loading the rest of the tables (like the fonttools subsetter does when --no-glyph-names option), and then you will no longer see the glyph names parsed from the post table, but those "made up" by FontTools based on a glyph's character mappings.

**Reply by justvanrossum** [reply]
> Do you know how GDI map the caracter to an glyph?
> I should have mentionned previously, but my goal is to reproduce how GDI map an caracter (like ë) to the glyph.

You'd have to look at the cmap to see what encoding it is mapping (if it's not unicode). So in this case you'd have to figure out that the only cmap available takes MacRoman. Then, you encode your input character as MacRoman, and check the ord of that in the subtable's cmap.

**Reply by anthrotype** [reply]
maybe you are also missing an important detail (not sure if was already mentioned earlier), that in the real cmap table (at the font binary level) the mappings are from codepoints to glyph _indices_ (or GIDs), but fontTools internally uses glyph _names_ (strings parsed from post/CFF or made up based on cmap) instead of GIDs. So you see the cmap as a dictionary keyed by integers (the codepoint in unicode or some other 8-bit encoding) and the values as glyph name strings, where _actually_ the cmap table (i.e. the one that GDI sees) only maps codepoints to glyph IDs. Does that make sense?

**Reply by moi15moi** [reply]
> Discard the glyph names altogether, by setting post to format 3.0 before loading the rest of the tables (like the fonttools subsetter does when --no-glyph-names option), and then you will no longer see the glyph names parsed from the post table, but those "made up" by FontTools based on a glyph's character mappings.

I am not sure if I understood.

I did that, but the codepoints from the cmap seems to be the same has before (so it seems to be influenced by the post table):
```py
from fontTools.ttLib.ttFont import TTFont

font = TTFont("test.ttf")
# From https://github.com/fonttools/fonttools/blob/97ed3a61cde03e17b8be36f866192fbd56f1d1a7/Lib/fontTools/subset/__init__.py#L2402
font["post"].formatType = 3.0
font.saveXML("test.xml")
```


> Does that make sense?

Yes, it is. There are many [cmap format](https://learn.microsoft.com/en-us/typography/opentype/spec/cmap), so I understand why fonttools works like that. It make it easier to interact with the cmap table.

What I wanna do seems to not be in the scope of fonttools.

**Reply by anthrotype** [reply]
the way the subsetter does is a bit of a hack, we should add support for this in ttLib...

https://github.com/fonttools/fonttools/blob/97ed3a61cde03e17b8be36f866192fbd56f1d1a7/Lib/fontTools/subset/__init__.py#L3519-L3535

**Reply by khaledhosny** [reply]
> I did that, but the codepoints from the cmap seems to be the same has before (so it seems to be influenced by the post table):
> 
> ```python
> from fontTools.ttLib.ttFont import TTFont
> 
> font = TTFont("test.ttf")
> # From https://github.com/fonttools/fonttools/blob/97ed3a61cde03e17b8be36f866192fbd56f1d1a7/Lib/fontTools/subset/__init__.py#L2402
> font["post"].formatType = 3.0
> font.saveXML("test.xml")
> ```

You need to save the font first then load it again so that FontTool synthesizes new glyph names.

**Reply by moi15moi** [reply]
> You need to save the font first then load it again so that FontTool synthesizes new glyph names.

This isn't possible in my use case :/
It would make the process way too slow and since I can load up to 5000 fonts, it would take an huge amount of space to "resave" the font.

**Reply by khaledhosny** [reply]
You don’t have to write the font to disk, you can use `io.BytesIO`

**Reply by moi15moi** [reply]
It doesn't seems to really change anything, because the caracter ``ë`` is still linked to the glyph ``ä``
```py
import io
from fontTools import ttLib
from fontTools.ttLib.ttFont import TTFont

font = TTFont("test.ttf",)
font["post"].formatType = 3.0

font_bytes = io.BytesIO()

font.save(font_bytes)
font.saveXML("before.xml")

font = TTFont(font_bytes)

font.saveXML("after.xml")
```

Before.xml:
```xml
...
<map code="0x98" name="edieresis"/>
...
 <glyf>
    <TTGlyph name="edieresis" xMin="31" yMin="10" xMax="565" yMax="769">
      <contour>
        <pt x="212" y="478" on="1"/>
        <pt x="237" y="485" on="0"/>
...
```

After.xml:
```xml
...
<map code="0x98" name="glyph00115"/>
...
<glyf>
    <TTGlyph name="glyph00115" xMin="31" yMin="10" xMax="565" yMax="769">
      <contour>
        <pt x="212" y="478" on="1"/>
        <pt x="237" y="485" on="0"/>
...
```

**Reply by justvanrossum** [reply]
You've got it backwards. Character ë is 0x91 in MacRoman. In the font, 0x91 is mapped to glyph ID 108. Which with the bad post table gets glyph *name* "adieresis", but the *shape* it references looks like the character ë. In the "after" xml dump, it is "glyph00108". Nothing really changed, except the glyph names, which aren't relevant for rendering. 

Your original question was "Does this font support character ë". This can be answered by your code, as long as you manage to convert ë to 0x91. For that question, the glyph name is not relevant either.

**Reply by moi15moi** [reply]
> This can be answered by your code, as long as you manage to convert ë to 0x91.

I thought all the thing with the glyph name would "magically" change the codepoints.
So to be clear, I have no idea how to convert ë to 0x91. That was my original question.

**Reply by justvanrossum** [reply]
```python
>>> s = "ë"
>>> ord(s.encode("macroman"))
145
>>> hex(_)
'0x91'
```

**Reply by moi15moi** [reply]
This conversation is getting ridiculous because i am not clear.

I want to know how to correct the codepoints in the cmap so that the character ë is well associated with the codepoint 0x91 (currently, it is associated with the codepoint 0x98).

**Reply by justvanrossum** [reply]
> I want to know how to correct the codepoints in the cmap so that the character ë is well associated with the codepoint 0x91 

When rendering, or when figuring out whether the font supports a certain (unicode) character, you do the following:

- Look for a unicode cmap subtable. If found, you're done (check if the code point exists in the cmap). This is not the case for this font because the font does not have a unicode cmap.
- Find the next best cmap, which **in this case** is a cmap subtable for **MacRoman**
- Now, if you want to render "ë", or want to know whether "ë" exists in this font, you will have to find the **MacRoman** code point for "ë". Which is 0x91.

So far so good.

> (currently, it is associated with the codepoint 0x98).

No it is not, you've still got it backwards.

The **character** ë (Unicode U+00EB, MacRoman 0x91) is mapped to **glyph id** 98 (decimal), by the MacRoman **cmap subtable**.

If you use **this font** to  **render** the glyph with **glyph id** 98 (decimal), you will see a shape that **correctly** represents the **character** ë.

If you look at this font **with fonttools**, you will notice that **glyph id** 98 (decimal) was assigned the **name** "adieresis". This **name** is incorrect, due to the **post** table being wrong.

- The **cmap** is correct
- The **glyph shapes** are correct
- The **names** are "wrong" because the post table is wrong.
- For rendering, or finding out whether a **character** exists in the font the **glyph names** are **irrelevant**
- The font **works correctly despite the post table being wrong**
- It just seems weird in fonttools, because it relies on a non-buggy post table to assign meaningful names
- Due to the **buggy post table** the names you see in fonttools **for this buggy font** are **meaningless**.

I really don't know how to explain it any clearer. I have now tried in 3, 4 different ways.

**Reply by justvanrossum** [reply]
- The glyphs inside a TTF are stored in an array in the `glyf` table
- Much simplified: font renderers find the glyph they want to render by looking up a character in the cmap. This gives them a glyph ID. The glyph ID is the index of the glyph into the glyphs array in the `glyf` table. They will render the glyph specified by that glyph ID.
- The `post` table has various ways to assign a name to each glyph ID (or none at all)
- So there often is a one-on-one mapping between glyph names and glyph IDs.
- Generally, as a font designer or engineer, it is much more pleasant to work with glyph names than with glyph IDs. The IDs differ per font, and you can not easily tell what the glyph may represent by looking at the glyph ID.
- Therefore, FontTools chooses to assign glyph names to each glyph ID, and uses them **instead of glyph IDs**, and it finds names by reading the `post` table
- Now, what if the `post` table contains an incorrect or weird glyph ID to glyph name mapping? Right: the names you see in FontTools now no longer make sense. This is what's happing in the font we're looking at.

**Reply by anthrotype** [reply]
Just's last two replies above should be "marked as answer".

**Reply by moi15moi** [reply, marked as answer]
Ok, I did not understood that the codepoints were already good, but it was only the name that wasn't.

I should have verified earlier that the glyph were already by creating an svg for each glyph and verify them:
```py
from fontTools.ttLib.ttFont import TTFont
from fontTools.pens.svgPathPen import SVGPathPen
from fontTools.pens.transformPen import TransformPen

font = TTFont("test.ttf")

glyphSet = font.getGlyphSet()

for table in font["cmap"].tables:
    for codepoint, name in table.cmap.items():

        svgpen = SVGPathPen(glyphSet)
        pen= TransformPen(svgpen , (1, 0, 0, -1, 0, 750))
        glyphSet[name].draw(pen)
        path = svgpen.getCommands()

        with open(f"{hex(codepoint)}.svg", "w") as text_file:
            text_file.write('<?xml version="1.0" encoding="UTF-8"?>\n')
            text_file.write('<svg baseProfile="full" width="1500" height="1500" xmlns="http://www.w3.org/2000/svg">\n')
            text_file.write(f'<path fill="black" d="{path}" />\n')
            text_file.write('</svg>')
```

So, in brief, here is what I was looking for:
```py
from fontTools.ttLib.ttFont import TTFont

def get_cmap_encoding(platformID, platEncID):
    encoding_map = {
        1: { # Mac
            0: 'mac_roman', # Need to verify if GDI has some exception with some langID
        },
        3: { # Microsoft
            0: 'utf_16_be',
            1: 'utf_16_be',
            2: 'shift_jis',
            3: 'gb2312',
            4: 'big5',
            5: 'euc_kr',
            6: 'johab',
            10: 'utf_16_be',
        },
    }

    return encoding_map.get(platformID, {}).get(platEncID, None)

def is_glyph_supported(font: TTFont, caracter: str):
    if len(caracter) != 1:
        raise Exception("Glyph need to be 1 character.")

    for table in font["cmap"].tables:
        cmap_encoding = get_cmap_encoding(table.platformID, table.platEncID)

        # Cmap isn't supported
        if cmap_encoding is None:
             continue
        
        if ord(caracter.encode(cmap_encoding)) in table.cmap:
            return True

    return False

font = TTFont("test.ttf")
print(is_glyph_supported(font, "ë"))
```
